xarray/core/variable.py

```diff
@@ -41,6 +41,18 @@
 # https://github.com/python/mypy/issues/224
 BASIC_INDEXING_TYPES = integer_types + (slice,)  # type: ignore
 
+VariableType = TypeVar("VariableType", bound="Variable")
```

Comment:
This looks in fine to me, though I think mypy would also be equally happy with the `bound="Variable"` part as long as you're doing it on a method.

Reply:
@shoyer I don't think I understand - could you elaborate?

Reply:
From mypy's perspective, it would be equivalent to write:
```
T = TypeVar('T')

class Variable:
  ...
  def some_method(self: T) -> T:
    ...
```

Reply:
Nope, wish it would.
```python
from typing import TypeVar

T = TypeVar("T")


class C:
    x: int

    def __init__(self, x: int):
        self.x = x

    def copy(self: T) -> T:
        return type(self)(self.x)
```
mypy 0.730 output:
```
13: error: Too many arguments for "object"
13: error: "T" has no attribute "x"
```

Reply:
Interesting. I've mostly been using Google's pytype for type checking, which handles this case.